### PR TITLE
Removed internal ids usage outside models

### PR DESCRIPTION
**Pull request by blondered**
- Removed `assume_external_ids` parameter in `recommend` and `recommend_to_items` model methods 
- Added `on_unsupported_targets` in `recommend` and `recommend_to_items` model methods 
- Added `filter_on_interactions_df_row_indexes` method of `Dataset`
- Fixed `IntraListDiversity` metric computation in `cross_validate`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,9 +11,20 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
 ### Fixed
 - `display()` method in `MetricsApp` ([#169](https://github.com/MobileTeleSystems/RecTools/pull/169))
+- `IntraListDiversity` metric computation in `cross_validate` ([#177](https://github.com/MobileTeleSystems/RecTools/pull/177))
+
+### Removed
+- [Breaking] `assume_external_ids` parameter in `recommend` and `recommend_to_items` model methods ([#177](https://github.com/MobileTeleSystems/RecTools/pull/177))
+
+
+### Fixed
```
Comment (feldlime) [inline CHANGELOG.md:25]: Duplicated 2 lines

File: rectools/dataset/dataset.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#  Copyright 2022-2024 MTS (Mobile Telesystems)
+#  Copyright 2022 MTS (Mobile Telesystems)
```
Comment (feldlime) [inline rectools/dataset/dataset.py:1]: Why? Did you use the script?

File: rectools/dataset/dataset.py [hunk 3]
```diff
@@ -245,3 +246,67 @@ def get_raw_interactions(self, include_weight: bool = True, include_datetime: bo
         pd.DataFrame
         """
         return self.interactions.to_external(self.user_id_map, self.item_id_map, include_weight, include_datetime)
+
+    def filter_on_interactions_df_row_indexes(
+        self,
+        row_indexes_to_keep: np.ndarray,
```
Comment (feldlime) [inline rectools/dataset/dataset.py:252]: Since we refused python 3.7 support, let's start adding types for numpy arrays. Please specify here that this is array of integers

File: rectools/model_selection/cross_validate.py [hunk 4]
```diff
@@ -1,71 +1,15 @@
-#  Copyright 2023-2024 MTS (Mobile Telesystems)
```
Comment (feldlime) [inline rectools/model_selection/cross_validate.py:1]: Why did you remove the license header?

File: rectools/model_selection/cross_validate.py [hunk 5]
```diff
@@ -132,34 +86,23 @@ def cross_validate(  # pylint: disable=too-many-locals
...
-        test_users = interactions_df_test[Columns.User].unique()  # 1x internal
-        catalog = interactions_df_train[Columns.Item].unique()  # 1x internal
-
-        if items_to_recommend is not None:
-            item_ids_to_recommend = dataset.item_id_map.convert_to_internal(
-                items_to_recommend, strict=False
-            )  # 1x internal
-        else:
-            item_ids_to_recommend = None
+        interactions_df_test = dataset.get_raw_interactions().iloc[test_ids]
```
Comment (feldlime) [inline rectools/model_selection/cross_validate.py:102]: Seems really ineffective:
1. You can get raw interactions once before the cycle.
2. We invented internal ids to avoid memory errors caused by big external ids, and here this logic fails.

What can be done now: you can use your new method `filter_on_interactions_df_row_indexes` to get test dataset, and then extract raw interactions from it. It makes sense since test part is usually much smaller than dataset. Yo can also add an argument to your method like `keep_features` and set it to `False` here to make it more effective.

What can we do in future: avoid converting anything to external ids and calculate metrics on internal ones.

File: rectools/models/base.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-#  Copyright 2022-2024 MTS (Mobile Telesystems)
+#  Copyright 2022 MTS (Mobile Telesystems)
```
Comment (feldlime) [inline rectools/models/base.py:1]: ?

File: tests/models/test_base.py [hunk 7]
```diff
@@ -461,16 +342,69 @@ def test_full_model_works_for_all_without_features(self, kind: str) -> None:
...
-        targets = self.colds[kind]
+    def test_not_cold_models_with_cold_targets(self, dataset_key: str, kind: str, model_key: str) -> None:
+        targets = self.colds[kind] + self.hots[kind]
+
+        # raise
         with pytest.raises(ValueError, match="doesn't support recommendations for cold"):
-            self._get_reco(targets, model_key, dataset_key, kind)
+            self._get_reco(targets, model_key, dataset_key, kind, on_unsupported_targets="raise")
+
+        # ignore
```
Comment (feldlime) [inline tests/models/test_base.py:352]: Please NEVER test multiple cases in one test.  Ask me if you want me to explain why.    
Please, either write multiple tests, or use `parametrize` or `subtest`

File: CHANGELOG.md (hunk 1 above)
Comment (blondered) [reply CHANGELOG.md]: Fixed

File: rectools/dataset/dataset.py (hunk 2 above)
Comment (blondered) [reply rectools/dataset/dataset.py]: Sorry. Fixed

File: rectools/dataset/dataset.py (hunk 3 above)
Comment (blondered) [reply rectools/dataset/dataset.py:252]: Decided to do it later

File: rectools/model_selection/cross_validate.py (hunk 4 above)
Comment (blondered) [reply rectools/model_selection/cross_validate.py]: Fixed

File: rectools/model_selection/cross_validate.py (hunk 5 above)
Comment (blondered) [reply rectools/model_selection/cross_validate.py]: I did the first one.
Can't see any optimization with calling `filter_on_interactions_df_row_indexes` for test interactions. It calls `get_raw_interactions` under the hood

File: rectools/models/base.py (hunk 6 above)
Comment (blondered) [reply rectools/models/base.py]: Fixed

File: tests/models/test_base.py (hunk 7 above)
Comment (blondered) [reply tests/models/test_base.py]: Fixed

File: rectools/model_selection/cross_validate.py (hunk 5 above)
Comment (feldlime) [reply rectools/model_selection/cross_validate.py]: First one was the obvious thought but it still not enough.  We can't keep raw interactions in memory.

Comment (feldlime) [reply rectools/model_selection/cross_validate.py]: > Can't see any optimization with calling filter_on_interactions_df_row_indexes for test interactions. It calls get_raw_interactions under the hood

Conclusion: we need to rewrite `filter_on_interactions_df_row_indexes`

Comment (blondered) [reply rectools/model_selection/cross_validate.py]: What do you want to do? get internal interactions.df, apply [test_ids] and then convert to external?

Comment (blondered) [reply rectools/model_selection/cross_validate.py]: Fixed

File: rectools/dataset/dataset.py [hunk 8]
```diff
@@ -245,3 +246,66 @@ def get_raw_interactions(self, include_weight: bool = True, include_datetime: bo
...
+            Otherwise internal -> 2x internal ids mapping will be created.
+        keep_features_for_removed_entities : bool, default `True`
+            Whether to keep all features for users and items that are not hot any more.
+
+        Returns
+        -------
+        Dataset
+            Filtered dataset that has only selected interactions, new ids mapping and processed features.
+        """
+        interactions_df = self.interactions.df.loc[row_indexes_to_keep]
```
Comment (feldlime) [inline rectools/dataset/dataset.py:276]: Why did you change `iloc` to `loc`? Sure it's safe? We don't pay attention to indices usually

Comment (feldlime) [reply rectools/dataset/dataset.py:276]: Btw you wrote in docstring `Original dataset interactions df row indexes` but we don't keep original indexes, so they are row numbers actually

File: rectools/model_selection/cross_validate.py [hunk 9]
```diff
@@ -132,34 +100,26 @@ def cross_validate(  # pylint: disable=too-many-locals
...
-        # We need to avoid fitting models on sparse matrices with all zero rows/columns =>
-        # => we need to create a fold dataset which contains only hot users and items for current training
-        fold_dataset = _gen_2x_internal_ids_dataset(
-            interactions_df_train, dataset.user_features, dataset.item_features, prefer_warm_inference_over_cold
+        fold_dataset = dataset.filter_on_interactions_df_row_indexes(
+            row_indexes_to_keep=train_ids,
+            keep_external_ids=True,
+            keep_features_for_removed_entities=prefer_warm_inference_over_cold,
         )
+        interactions_df_test = dataset.interactions.df.loc[test_ids]
```
Comment (feldlime) [inline rectools/model_selection/cross_validate.py:116]: Why don't you use your new dataset method here?

File: rectools/dataset/dataset.py (hunk 8 above)
Comment (blondered) [reply rectools/dataset/dataset.py:276]: `filter_on_interactions_df_row_numbers`?

Changed to iloc back

Comment (feldlime) [reply rectools/dataset/dataset.py:276]: I'd name it shorter since it's a public method. E.g. `iloc_interactions`

File: rectools/model_selection/cross_validate.py (hunk 9 above)
Comment (feldlime) [reply rectools/model_selection/cross_validate.py:116]: But it'd be just one line of code instead of 3, no?

Comment (blondered) [reply rectools/model_selection/cross_validate.py:116]: It would be 2 lines here instead of 3.
5 lines in the func itself  and 20 lines for tests :)
right now the logic is clear and all is fine. no need for dataset when we need  just interactions

File: rectools/dataset/dataset.py (hunk 8 above)
Comment (blondered) [reply rectools/dataset/dataset.py:276]: That's not good since we'll get a dataset from this method (not interactions). And we will write more filters for dataset in the future (like filter on features selected)
so `filter_on` beginning is very reasonable

Comment (feldlime) [reply rectools/dataset/dataset.py:276]: ok, then let's just name it `filter` (I don't really understand `on`)

File: rectools/model_selection/cross_validate.py (hunk 9 above)
Comment (feldlime) [reply rectools/model_selection/cross_validate.py:116]: Didn't get your point, sorry. 
You have the func already, just call it and take interactions.  
It'd be more readable than doing manual manipulations.  
Why do you need add lines to the func?

Comment (blondered) [reply rectools/model_selection/cross_validate.py:116]: It's not appropriate since new func is for dataset creation. It has more going on than just converting ids of interactions to external ones. For example it will construct new ids maps. We just don't need it here

File: rectools/dataset/dataset.py (hunk 8 above)
Comment (blondered) [reply rectools/dataset/dataset.py]: ok)
renamed to `filter_interactions_df_rows`

Comment (feldlime) [reply rectools/dataset/dataset.py]: You said we're going to filter features with this method in the future - let's then call it just `filter`?

Comment (blondered) [reply rectools/dataset/dataset.py]: I thought there will be different `filter_<on something>` methods. 
I'm not sure if we want to do all filters in the same method. It's probably a good idea - this way we never need to call a chain of filters (which is inefficient). So you're probably right.

It's gonna be a `filter` method. And possible arguments might be (just my imagination): `interactions_df_rows`, `user_features_subset`, `item_features_subset` and so on (if any). Right?

Comment (feldlime) [reply rectools/dataset/dataset.py]: yes, sounds reasonable

Comment (blondered) [reply rectools/dataset/dataset.py]: Fixed

**Review by feldlime — APPROVED** [review]
